pkg/endpoint/policy.go

```diff
@@ -70,10 +70,6 @@ type RegenerationContext struct {
 	// used to generate useful log messages.
 	Reason string
 
-	// ReloadDatapath forces the datapath programs to be reloaded. It does
-	// not guarantee recompilation of the programs.
-	ReloadDatapath bool
```

Comment:
I think this field should be contained within `RegenerationContext` like it is currently; it allows us to not have to update function signatures / leaking of implementation, and instead store state relevant to a regeneration inside of a type. 

The problem is that in `RegenerateAllEndpoints`, we create a new `RegenerationContext` using only the `reason` of the `RegenerationContext` that was passed in:

https://github.com/cilium/cilium/blob/0a17787e8047092730446441189ed656b55654d3/pkg/endpointmanager/manager.go#L281-L308

We can create a copy of this `RegenerationContext` instead and keep the change localized to that function.

Reply:
@ianvernon I know, I've decided to make it explicit because RegenerationContext should be used for statistics, as a parameter you will have to opt-in either you want to reload the datapath or not. Next time we will add a field to `RegenerationContext` and if we don't update the constructor we will have the same problem.

Reply:
Can you explain your rationale as to why `RegenerationContext` only be used for statistics? If we create a copy, it can be mutable for a regeneration for a given endpoint. We don't call `endpoint.Regenerate` in many places, so I think we can control for this. 

I think it's fine to ensure that we create always create a copy of a `RegenerationContext` before passing it down to `endpoint.Regenerate`. This is necessary if we want to add more state into the type, which I am doing in #6191 .

Reply:
@ianvernon actually the `RegenerationContext` should only provide context to regenerate() calls to determine the caller. It should not even have the stats there. `RegenerationContext` is our context.Context implementation. We should not mix things.

Reply:
This may be partly my (mis?)interpretation. I was taking the `RegenerationContext` as the general context that is necessary for regeneration, ie in the long term it could grow to include all parameters that are necessary for regeneration and this structure would encapsulate that information.

Reply:
I was the impetus for making this a parameter (also the reason the PR that broke this happened! I am all the faults). My concern was that it was too easy to accidentally repeat this sort of mistake. In this case reloadDatapath is used only one level down (and doesn't seem to be used in other places the context is passed).
In any case, my preference is to force a copier of a regenerationContext to decide if this parameter should be passed on as-is or not. An alternative would be to make the `newRegererationContext()` constructor also take a bool argument. It could also take a pointer and do the copies internally, but I don't know how often we change the internal members of the context struct (apart from stats, not at all today, I think).
I'm less keen on doing a direct copy (e.g. `newCtx := *oldCtxPtr`) unless we also remove the constructor. It would be very likely for someone to copy-paste the one line and miss the copy-assignment elsewhere in the function.

Reply:
I'll approve this since it solves the issue and we need to backport it. But, I'd like to continue this discussion.